### DISCUSSION TITLE
App Router Authorization and Authentication (Storing The Token)

### DISCUSSION THREAD

**Discussion by samirj12** [vercel/next.js#71801] [Help]
### Summary

I'm encountering issues with authentication and authorization in my Next.js 14 App Router project. The login page is a server component with a client-side login form. After a successful login, the server returns an access token and refresh token.
Given that the app is a dashboard relying heavily on client-side components, I'm unsure whether to store tokens client-side (in local storage) or server-side (in cookies).

1. **Token Storage**: For each request, the authorization header needs the access token. Considering security, is client-side or server-side storage preferable?
2. **Route Protection**: I need a middleware to restrict access to certain routes based on user roles. For example, basic users shouldn't access `/admin` pages. How can I achieve this without accessing local storage in the middleware?
3. **Function-Level Authorization**: Beyond route-level permissions, I need to verify if a user has specific permissions within a route, like deleting a repository. How can I enforce this?
4. **Token Refresh**: To handle token expiration, I want to automatically request a new token on 401 errors and retry the failed request. How can I implement this efficiently?

**Comment by sundargautam18** [comment]
Here is the best way: Here is the better and simple solution to implement this : https://medium.com/@sundargautam2022/implementing-refresh-token-with-nextjs-15-using-app-router-with-cross-api-different-api-5682f83f9802
